### PR TITLE
Use scheduler cache in affinity priority functions

### LINKED ISSUE #79629: Scheduler: affinity priority function lists nodes from the informer cache instead of scheduler cache
**What happened**:
#79076 modified the scheduler configuration such that NodeInfo is read from the scheduler cache. [NodeInfo is used in calculating pod affinity priority along with NodeLister](https://github.com/kubernetes/kubernetes/blob/38e8b8a9f65e66b779a0985e17868b036c1d601b/pkg/scheduler/algorithmprovider/defaults/register_priorities.go#L73). NodeLister is a light wrapper around a corelisters.NodeLister, which reads from the informer cache.

Listing nodes from one cache, while reading their info from another can potentially lead to issues similar to the ones described in #73956 

**What you expected to happen**:
Predicates and priority functions should read from the scheduler cache for consistency.

/kind bug
/sig scheduling
/help

### PR DESCRIPTION
**Pull request by alculquicondor**
Make the cache implement NodeLister and expose it to the priority
functions. This way, the priority functions make use of a single cache,
the scheduler's, instead of mixing it with the lister's caches.

Fixes #79629

/kind cleanup

```release-note
NONE
```

### CODE REVIEW

File: pkg/scheduler/factory/factory.go
Comment (draveness) [inline pkg/scheduler/factory/factory.go]: why move theses functions here?

File: pkg/scheduler/factory/factory.go
Comment (draveness) [inline pkg/scheduler/factory/factory.go]: same here

File: pkg/scheduler/internal/cache/cache.go
Comment (draveness) [inline pkg/scheduler/internal/cache/cache.go]: make([]*v1.Node, len(cache.Nodes)

File: pkg/scheduler/internal/cache/interface.go
Comment (draveness) [inline pkg/scheduler/internal/cache/interface.go]: nit: please revert this line.

**Review by draveness — CHANGES_REQUESTED** [review]

File: pkg/scheduler/factory/factory.go
Comment (alculquicondor) [reply pkg/scheduler/factory/factory.go]: To keep them visible in preparation for the TODO.

File: pkg/scheduler/factory/factory.go
Comment (alculquicondor) [reply pkg/scheduler/factory/factory.go]: To keep them visible in preparation for the TODO.

File: pkg/scheduler/internal/cache/cache.go
Comment (alculquicondor) [reply pkg/scheduler/internal/cache/cache.go]: Done

File: pkg/scheduler/internal/cache/interface.go
Comment (alculquicondor) [reply pkg/scheduler/internal/cache/interface.go]: Done, but there should be no need for that name.

File: pkg/scheduler/internal/cache/cache.go
Comment (alculquicondor) [reply pkg/scheduler/internal/cache/cache.go]: Done

File: pkg/scheduler/factory/factory.go
Comment (draveness) [reply pkg/scheduler/factory/factory.go]: Why not add a TODO before line 168?

**Review by draveness — CHANGES_REQUESTED** [review]

File: pkg/scheduler/factory/factory.go
Comment (draveness) [inline pkg/scheduler/factory/factory.go]: Why not Add a TODO here?

**Review by wgliang — CHANGES_REQUESTED** [review]

File: pkg/scheduler/internal/cache/cache.go
Comment (alculquicondor) [reply pkg/scheduler/internal/cache/cache.go]: any reason? quoted might be more readable.

File: pkg/scheduler/algorithm/types.go
Comment (alculquicondor) [reply pkg/scheduler/algorithm/types.go]: The problem is that List conflicts with PodLister's

File: pkg/scheduler/factory/factory.go
Comment (draveness) [inline pkg/scheduler/factory/factory.go]: move these fields to line 191 is not  necessary to me.

File: pkg/scheduler/factory/factory.go
Comment (draveness) [inline pkg/scheduler/factory/factory.go]: move these fields to line155 is not necessary to me.

**Comment by alculquicondor** [comment]
/assign @huang-wei
Tests are passing

File: pkg/scheduler/factory/factory.go
Comment (alculquicondor) [reply pkg/scheduler/factory/factory.go]: I'll leave this to the approver.

File: pkg/scheduler/factory/factory.go
Comment (alculquicondor) [reply pkg/scheduler/factory/factory.go]: Leaving to the approver.

**Review by draveness — CHANGES_REQUESTED** [review]
Please squash the commits into one

**Comment by alculquicondor** [comment]
I'll let the bot handle the squash during merge. Having a history is useful for reviews.

**Comment by draveness** [comment]
> I'll let the bot handle the squash during merge. Having a history is useful for reviews.

The first three commits have failed the tests, make the commits atomic is preferable in the codebase.

File: pkg/scheduler/core/generic_scheduler.go
Comment (alculquicondor) [reply pkg/scheduler/core/generic_scheduler.go]: I will leave this to a separate PR. Even though the change is trivial in production code, several tests get affected.

Comment (ahg-g) [reply pkg/scheduler/core/generic_scheduler.go]: ok, please create an issue so we can track it.

/lgtm

Comment (alculquicondor) [reply pkg/scheduler/core/generic_scheduler.go]: Done: #80322

File: pkg/scheduler/factory/factory.go
Comment (bsalamat) [reply pkg/scheduler/factory/factory.go]: I agree with @draveness. Why are we moving these lines?

File: pkg/scheduler/internal/cache/interface.go
Comment (bsalamat) [inline pkg/scheduler/internal/cache/interface.go]: our cache interface should not depend on `predicates`. Other components such as `predicates` are users of `Cache`. It makes sense for them to depend on `Cache`, not the other way round. In fact, we should remove the dependency on `algorithm` that I added in the past. Please move the types you need from `predicates` to here if possible.

File: pkg/scheduler/internal/cache/interface.go
Comment (draveness) [inline pkg/scheduler/internal/cache/interface.go]: Why are we use the interface instead of method signatures here? We could revert this change to solve @bsalamat's comment [here](https://github.com/kubernetes/kubernetes/pull/80220/files#r305156226)

File: pkg/scheduler/factory/factory.go
Comment (alculquicondor) [reply pkg/scheduler/factory/factory.go]: Done

File: pkg/scheduler/internal/cache/interface.go
Comment (alculquicondor) [reply pkg/scheduler/internal/cache/interface.go]: A signature should be written once. We need to give more thought to this.

File: pkg/scheduler/internal/cache/interface.go
Comment (alculquicondor) [reply pkg/scheduler/internal/cache/interface.go]: It seems that nor `predicates` nor `algorithms` currently depend on Cache. And I don't think it's a good idea to make them do so. Maybe all these interfaces should go to a brand new package `pkg/scheduler/listers`. If you think that's a good idea, we should do that in a separate PR and leave this PR with the embedded interfaces.

Comment (bsalamat) [reply pkg/scheduler/internal/cache/interface.go]: I know that there is no direct dependency on `cache` in `predicates` or `algorithms`, but they actually use data from cache. `NodeInfo` which is heavily used in `predicates` and `priorities` is coming from cache. In fact, we wanted to move `NodeInfo` into our `cache` directory, but since kubelet uses it, we couldn't put in under our `internal` package.
We can create a separate package for listers at `pkg/scheduler/internal/listers`. Whether you want to do it in this PR or a follow-up is up to you, but please do not create dependency in cache on `predicates`.

Comment (alculquicondor) [reply pkg/scheduler/internal/cache/interface.go]: Reverting. However, we still to move `NodeInfo`, without it being in an internal package.

I created #80372 to track this.

File: pkg/scheduler/factory/factory.go
Comment (draveness) [inline pkg/scheduler/factory/factory.go]: This needs to be addressed according to @bsalamat's comment

Comment (alculquicondor) [reply pkg/scheduler/factory/factory.go]: Not necessarilly, that referred to a different struct.

**Review by bsalamat — COMMENTED** [review]
/lgtm
/approve

Thanks, @alculquicondor!

**Comment by alculquicondor** [comment]
/hold

I just realized that the custom lister used on tests needs to go away in favor of the cache.

**Comment by alculquicondor** [comment]
/hold cancel

PTAL once tests pass

**Pull request merged**
